### PR TITLE
Meili/orchestration thoughts

### PR DESCRIPTION
**Pull request by nokuyo**

### CODE REVIEW

File: mkdocs/docs/assets/orchestration.png
[the remark this answers is not in the document]
Comment (bens-schreiber) [reply mkdocs/docs/assets/orchestration.png:1]: also, wrangler.toml is a root level user defined file that we just augment _if required_, not generate

File: mkdocs/docs/meili/v0.0.3/orchestration_thoughts.md [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+## Design Philosophies
+
+### 1) Reproducible and hermetic
+Reproducible is a given since Cloesce's output is deterministic. What we really need to implement is making sure npm orchestration is hermetic. Builds shouldn't depend on the host machine's quirks and no matter what OS, PATH variables, global installs, when we install the package on a random device, ideally everything should behave the same.
+
+### 2) Cross-platform
+Goes hand in hand with 1, it should work across macOS/Linux/Windows without any weird shell quirks
+
+### 3) Speed
+We should aim for orchestration to be reasonably fast. Usually npm packages will cache things like configuration flags, model files if not changed, things like that so after the first run the npm command will run significantly faster, and since one of the big use cases for Cloesce is to advance vibe coding beyond just localhost, I assume users will be running cloesce over and over.
```
Comment (bens-schreiber) [inline mkdocs/docs/meili/v0.0.3/orchestration_thoughts.md:14]: absolutely, post MVP I want to explore shifting Cloesce to be an incremental compiler for maximum build speed in development

File: mkdocs/docs/meili/v0.0.3/orchestration_thoughts.md [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+## Commands and Architecture
+
+The first thing I want to cover is commands. Ideally we run `cloesce run` and everything is generated, all the user has to do is `wrangler run dev`
+
+But what goes under the hood? First thing I want to make clear is that npm should not just be a underlying bash script. We should treat the build as a dataflow, not a bunch of shell commands. **Essentially we need to build a directed acyclic graph since most good npm packages use frameworks or have their own systems that do this.** Each step (TS to IR, IR to workers.ts, IR to SQL, Wrangler build, etc.) is a task node which consumes inputs and produces outputs. So the edges say "the output of this task node will feed into the next task", and when we rebuild we can detect what the user changed, find only the tasks downstream that need to be rerun, and rebuild the minimal set. This means that things will be way faster when we rebuild, we can run things via parallelism if we want to (we can generate workers AND d1 at the same time), and we have observability into what changed for logging purposes.
```
Comment (bens-schreiber) [inline mkdocs/docs/meili/v0.0.3/orchestration_thoughts.md:33]: This is the heart of incremental compilation, which I've been exploring. A naive solution is to create a merkle tree of the CIDL and DFS to find hash differences, but it will require a major rework of how our compiler works.

File: mkdocs/docs/meili/v0.0.3/orchestration_thoughts.md [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+So how do we have a unified binary then? We quickly run into issues for the average user when we consider that they might not have Rust installed, and different OS and CPU combos can create weird execution with binaries, overall does not sound fun to implement and test. This is where **WebAssembly, or WASM** comes in. WASM gives us a portable, precompiled binary that runs on any platform inside a Node process. So we can just compile the Rust generator into a `.wasm` file, then Node will run it via JS wrapper within the npm package, and boom, we have complete coverage over every OS and CPU. Now if a user wants to develop Tinder for Horses on Yocto Linux running on secret technology forbidden Uzbekistani 128 bit CPU architecture, they can do so and we don't have to test and fix random edge cases or juggle messy testing pipelines. Overall the set up seems pretty easy, just need to compile the rust generator on wasm and we're good to go, our JS wrapper will handle the rest. 
+
+And that's orchestration in a nutshell. Obviously up to change as I implement this but in terms of npm orchestration, this is **usually** how orchestration is done and it's what I intend to implement.
```
Comment (bens-schreiber) [inline mkdocs/docs/meili/v0.0.3/orchestration_thoughts.md:53]: It seems like the main focus of this initial issue on NPM orchestration is:
1. NPM and Rust to WASM production pipeline (builds publish an NPM package which houses our generator WASM)
2. After including the `cloesce` package to a TS project, running `npm cloesce run` should extract + generate
3. After running `npm cloesce run`, one should either get errors, or be able to just run `wrangler run dev`
4. A super naive setup to just run `npm cloesce run` on any change anywhere in the models dir (wrangler already has this somewhere, I'm sure we can tap into it)

Can you touch on the that first point a bit more? I'm totally unfamiliar with publishing NPM packages, and I haven't ever made a production build pipeline with GitHub actions before.

File: mkdocs/docs/meili/v0.0.3/orchestration_thoughts.md [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+Implementation wise, we need to make sure each run behaves the same and reads one source of truth, every file should be generated into the same `.generated/` file. Then we need to do content hashing and env fingerprints, meaning we read the bytes of important files like the TS models and hash them. This ensures that when we rebuild and the hash is the same, we can quickly keep going and build since we know nothing has changed. Then we formalize each build step as a worker with a clear contract, as in each step should have dependencies that must be completed before we run anything, what counts as inputs, the output it promises to write, and how we run everything. Then we build the DAG and do topological sort, we build the dependency graph from the registry and compute a safe execution order, skipping cache hits and executing everything in one go.
```
Comment (bens-schreiber) [inline mkdocs/docs/meili/v0.0.3/orchestration_thoughts.md:41]: for the sake of this PR we should stay simple: save a hash, if anything is changed at all just re-run the entire generate command. no dependencies, same execution order as defined in e2e tests, always ran the same way

File: mkdocs/docs/meili/v0.0.3/orchestration_thoughts.md [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+
+In terms of the testing pipeline, it seems like we can just do what we're doing already with our current testing pipeline, only that we use npm pack to locally simulate the published tarball and then use npm publish via github actions *iff* the run is successful and merged into main. We can use `npm version patch|minor|major` to indicate what version of Cloesce is being published as well.
```
Comment (bens-schreiber) [inline mkdocs/docs/meili/v0.0.3/orchestration_thoughts.md:53]: note this isn't a testing pipeline but a production pipeline

**Review by bens-schreiber — APPROVED** [review]

File: mkdocs/docs/meili/v0.0.3/orchestration_thoughts.md (hunk 5 above)
Comment (bens-schreiber) [reply mkdocs/docs/meili/v0.0.3/orchestration_thoughts.md:53]: I wonder if we even do want this to run on every merge, maybe we should have direct control of that and publish deliberate manual releases

Comment (nokuyo) [reply mkdocs/docs/meili/v0.0.3/orchestration_thoughts.md:53]: For official open source projects github as a *release* feature that we can probably integrate, either that or we have a staging branch and a producton branch. See https://github.com/systemd/systemd as an example
